File: PYTHON_SOLUTIONS/227.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s:
            return 0
        s = s.replace(" ", "")
        l,r = 0,1
        sign = 1 
        stack = []
        while l < len(s):
            if s[l].isdigit():
                while r < len(s) and s[r].isdigit():
                    r += 1
                num = int(s[l:r])
                stack.append(num * sign)
            elif s[l] == "+":
                sign = 1
            elif s[l] == "-":
                sign = -1
            else:
                op = s[l]
                lv = stack.pop()
                l += 1
                r = l + 1
                while r < len(s) and s[r].isdigit():
                    r += 1
                num = int(s[l:r])
                #print(lv, op, num)
                if op == "*":
                    stack.append(lv * num)
                else:
                    stack.append(abs(lv) / num * (1 if lv > 0 else -1))
            l = r
            r += 1
            #print(stack)
        return sum(stack)
                
            
        

File: CPP_SOLUTIONS/227.cpp
class Solution {
public:
    int calculate(string s) {
        int l = 0;
        int sign = 1, last = 1;
        stack<int> st;
        string tmp;
        for(char c: s) if(c != ' ') tmp += c;
        s = tmp;
        while(l < s.size()) {
            int r = l + 1;
            if(s[l] == '+') sign = 1 * sign;
            else if(s[l] == '-') sign = -1 * sign;
            else if(s[l] == '*') sign = 2;
            else if(s[l] == '/') sign = 3;
            else {
                while(r < s.size() && s[r] >= '0' && s[r] <= '9') r ++;
                int term = atoi(s.substr(l, r - l).c_str());
                if(sign > 0) {
                    if(sign == 1) st.push(term);
                    else if(sign == 2) {
                        int t1 = st.top();
                        st.pop();
                        st.push(t1 * term);
                    }
                    else  {
                        int t1 = st.top();
                        st.pop();
                        st.push(t1 / term);
                    }
                } else {
                    if(sign == -1) st.push(-term);
                    else if(sign == -2) {
                        int t1 = st.top();
                        st.pop();
                        st.push(-t1 * term);
                    }
                    else  {
                        int t1 = st.top();
                        st.pop();
                        st.push(-t1 / term);
                    }
                }
                sign = 1;
            }
            l = r;
        }
        int res = 0;
        while(!st.empty()) {
            res += st.top();
            st.pop();
        }
        return res;
    }
};
